### DISCUSSION TITLE
Next 12 ' TypeError: Class constructors cannot be invoked without 'new' '

### DISCUSSION THREAD

**Discussion by 0xmovses** [vercel/next.js#31135] [Help]
### What example does this report relate to?

es6 

### What version of Next.js are you using?

12.0.1

### What version of Node.js are you using?

16.10.0

### What browser are you using?

Chrome, Brave

### What operating system are you using?

macOS

### How are you deploying your application?

next start 

### Describe the Bug

Next can't compile this magic-link auth function. I have to wrap it in an if statement as magic needs access to the window object. 

After running yarn dev I get the error: 

```TypeError: Class constructors cannot be invoked without 'new'```

```
const [magic, setMagic] = useState<any>();
    const [AUTHORIZATION_FUNCTION, setAuthorizationFunction ] = useState<any>();

    if(typeof window !== "undefined"){
        const magic = new Magic(process.env.NEXT_PUBLIC_MAGIC, {
            extensions: [
                new FlowExtension({
                  rpcUrl: 'https://access-testnet.onflow.org',
                }),
              ],
          });

          const AUTHORIZATION_FUNCTION = magic.flow.authorization;

          setMagic(magic);
          setAuthorizationFunction(AUTHORIZATION_FUNCTION)
    }
```

my next.config.js file is set to default, no custom changes. 

My tsconfig.json: 

```
{
  "compilerOptions": {
    "target": "esnext",
    "lib": [
      "dom",
      "dom.iterable",
      "esnext"
    ],
    "allowJs": true,
    "skipLibCheck": true,
    "strict": true,
    "strictNullChecks": true,
    "forceConsistentCasingInFileNames": true,
    "noEmit": true,
    "esModuleInterop": true,
    "module": "esnext",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "jsx": "preserve",
    "incremental": true
  },
  "include": [
    "next-env.d.ts",
    "**/*.ts",
    "**/*.tsx",
    "desc.d.ts"
  ],
  "exclude": [
    "node_modules"
  ]
}
```

### Expected Behavior

The app should compile and run in the browser

### To Reproduce

clone the repo on branch melkonian/magiclink

https://github.com/INFLOWMusic/inflow-app

run yarn install or npm install

**Comment by timneutkens** [comment]
It seems you didn't follow the example exactly, the `new Magic` part is called in a separate function, not during render: https://github.com/vercel/next.js/blob/canary/examples/with-magic/pages/login.js#L14-L45

I'll convert this issue to discussions given that the example looks fine.

**Reply by 0xmovses** [reply]
Thanks Tim, I did actually write the code as you describe originally, in a separate function, but I got an error ``window is not defined`` as the magic link library makes use of the window object. This is why I then put it in this If block, puttting it inside a useEffect, gives the same error. 

However, I haven't tried it exactly as its written in the snippet you've linked. I'll give it a go! thanks

**Reply by 0xmovses** [reply]
Hmm I used the example project above above, Magic Link login works fine but when I add the FlowExtension I get this new error, ```TypeError: Cannot read properties of undefined (reading 'Internal')```

```
const magic = new Magic(process.env.NEXT_PUBLIC_MAGIC_PUBLISHABLE_KEY, {
        extensions: [
          new FlowExtension({
            rpcUrl: 'https://access-testnet.onflow.org',
          }),
        ],
      });
```

further more the Magic extension for the Flow Block chain has only 6 weekly downloads and was published a year ago https://www.npmjs.com/package/@magic-ext/flow . I've alerted the Magic Link dev team about this.
